src/javascript/app_2/Stores/base_store.js

```diff
@@ -0,0 +1,139 @@
+        session_storage_properties = null
+    ) {
+
+        Object.defineProperty(this, 'root_store', {
+            enumerable: false,
+            writable  : true,
+        });
+        Object.defineProperty(this, 'local_storage_properties', {
+            enumerable: false,
+            writable  : true,
+        });
+        Object.defineProperty(this, 'session_storage_properties', {
+            enumerable: false,
+            writable  : true,
+        });
+
+        this.root_store = root_store;
+        this.local_storage_properties   = local_storage_properties || [];
+        this.session_storage_properties = session_storage_properties || [];
+        
+        this.setupReactionForLocalStorage();
+        this.setupReactionForSessionStorage();
+        this.retrieveFromStorage();
+    }
+
+    /**
+     * Returns an snapshot to the current store
+     *
+     * @param {String[]} properties - A list of properties' names that should be in the snapshot.
+     *
+     * @return {Object} Returns a clone object for the store.
+     */
+    getSnapshot(properties) {
+        let snapshot = toJS(this);
+
+        if (!isEmptyObject(this.root_store)) {
+            snapshot.root_store = this.root_store;
+        }
+
+        if (properties && properties.length) {
+            snapshot = properties.reduce(
+                (result, p) => Object.assign(result, { [p]: snapshot[p]}),
+                {}
+            );
+        }
+
+        return snapshot;
+    }
+
+    /**
+     * Setups a reaction on properties which are mentioned in `local_storage_properties`
+     *  and invoke `saveToStorage` when there are any changes on them.
+     *
+     */
+    setupReactionForLocalStorage() {
+        if (this.local_storage_properties.length) {
+            reaction(
+                () => this.local_storage_properties.map(i => this[i]),
+                () => this.saveToStorage(this.local_storage_properties, BaseStore.STORAGES.LOCAL_STORAGE)
+            );
+        }
+    }
+
+    /**
+     * Setups a reaction on properties which are mentioned in `session_storage_properties`
+     *  and invoke `saveToStorage` when there are any changes on them.
+     *
+     */
+    setupReactionForSessionStorage() {
+        if (this.session_storage_properties.length) {
+            reaction(
+                () => this.local_storage_properties.map(i => this[i]),
+                () => this.saveToStorage(this.session_storage_properties, BaseStore.STORAGES.SESSION_STORAGE)
+            );
+        }
+    }
+
+    /**
+     * Saves the snapshot of the store includes properties that are mentioned in
+     *  first parameter in the storage that is mentioned in second parameters.
+     *
+     * @param {String[]} properties - A list of name of properties of the store  which they should be saved on the storage.
+     * @param {Symbol}   storage    - A symbol object that define the storage which the snapshot should stored on it.
```

Comment:
`A symbol object that defines the storage which the snapshot should be stored in`